### DISCUSSION TITLE
How to test render custom tag on single post ?

### DISCUSSION THREAD

**Discussion by dimaslanjaka** [hexojs/hexo#5203] [Question / Answer] [answered]
I was created my plugin custom shortcode.

how to test render single post with plugin enabled ?

i was used
```ts
process.cwd = () => __dirname;

import Hexo from 'hexo';
import path from 'path';

const hexo = new Hexo(__dirname);
async function main() {
  await hexo.init();
  await hexo.load(() => {
    //
  });
  const post = await hexo.render.render({ path: path.join(__dirname, 'source/lang/php.md') });
  console.log(post);
}

main();
```

but the result not render the shortcode tag, how to do it?

![image](https://github.com/hexojs/hexo/assets/12471057/58ea6285-8d6c-41ce-a7b1-712d1f3e3eb4)

but my plugin is working when i call `hexo generate` or `hexo server`

**Comment by lorezyra** [comment]
Shouldn't you specify the engine in the render call?

**Reply by dimaslanjaka** [reply]
ill try

**Reply by dimaslanjaka** [reply]
![image](https://github.com/hexojs/hexo/assets/12471057/d3fa542d-ffc9-4b1d-a3c1-47a57dec4091)
not working too

**Reply by dimaslanjaka** [reply]
![image](https://github.com/hexojs/hexo/assets/12471057/7aaa69f8-717e-4c38-9ddb-ffe063ade93b)

**Comment by dimaslanjaka** [comment]
oh maybe need load plugin like https://github.com/hexojs/hexo/blob/master/test/scripts/tags/blockquote.js

**Comment by dimaslanjaka** [comment, marked as answer]
solved using `hexo.loadPlugin` and `hexo.post.render`.

```ts
import Hexo from 'hexo';
import path from 'path';

const hexo = new Hexo(__dirname);
async function main() {
  await hexo.init().then(() => hexo.loadPlugin(require.resolve('hexo-shortcodes')));
  await hexo.load(async () => {
    const post = await hexo.post.render(path.join(__dirname, 'source/lang/php.md'));
    console.log(post);
  });
}

main();
```

**Reply by lorezyra** [reply]
Usually, I do the import from the top of the script and specify the path if it's not in the standard location. However, this doesn't look like you are building a tag/helper plugin for HEXO.

**Reply by dimaslanjaka** [reply]
> Usually, I do the import from the top of the script and specify the path if it's not in the standard location. However, this doesn't look like you are building a tag/helper plugin for HEXO.

look at here where i forked https://github.com/hexojs/hexo/blob/master/test/scripts/tags/blockquote.js#LL8C16-L8C96
look at here where i read the docs https://github.com/dimaslanjaka/hexo/blob/cfd7acbe0640d7c406c9d2d39e8aa7164786fd5a/lib/hexo/index.ts#L373-L381
here my custom shortcodes tag for hexo https://github.com/dimaslanjaka/hexo-shortcodes
here the result all of tags https://www.webmanajemen.com/docs/hexo-shortcodes/

**Reply by dimaslanjaka** [reply]
i doing test just for coverage using jest and coveralls. but all tags are working perfectly on deployment

**Reply by dimaslanjaka** [reply]
and test result below
![image](https://github.com/hexojs/hexo/assets/12471057/b4edd0a3-79d7-4aed-b1fe-daa7e54e99ab)

And all the shortcodes work fine and according to what I want, just need the post body (post.content) to validate shortcode result using JSDOM or cheerio

**Reply by dimaslanjaka** [reply]
btw I'm using a monorepo with a yarn workspace, which has a site documentation builder as well as separate test scopes.
